Two types game.

Stan-Kudri opened the pull request.

Stan-Kudri committed:
Add and Changed:

-Add interface "ISnake".
-Add two classes "SnakePassingBorders"(Past class "Snake") and "SnakeNotPassingBorders"(Past class "Snake", but three changed methods).

Stan-Kudri committed:
Added snake type selection.

blowin commented on Snake/ComponentsGame/GameMap.cs:
```diff
@@ -10,15 +13,28 @@ public class GameMap
-        private Snake _snake;
 
-        public GameMap(int width, int height, int snakeLength = 10)
+        public GameMap(int width, int height, TypeSnake type, int snakeLength = 10)
```
почему бы не передавать сразу ISnake?

blowin replied:
сделай какой-нибудь метод статический в Snake, который вернет тебе змейку по типу

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/ISnake.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public interface ISnake
+    {
+        public const char SymbolSnake = 'О';
```
этого тут не должно быть

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/SnakeNotPassingBorders.cs:
```diff
@@ -0,0 +1,115 @@
+
+namespace GameSnake.ComponentsGame.ItemGameMap.SnakeType
+{
+    public class SnakeNotPassingBorders : ISnake
```
посмотри можешь ли ты сделать базовый класс, думаю тут будет много дублирования

blowin commented on Snake/Program.cs:
```diff
@@ -22,6 +24,32 @@ static void WindowSetting(int width, int height)
+        var key = Console.ReadKey(true);
+        if (int.TryParse(key.KeyChar.ToString(), out var numberType))
+        {
+            if (numberType == 1)
```
https://learn.microsoft.com/en-us/dotnet/api/system.enum.tryparse?source=recommendations&view=net-7.0

Stan-Kudri committed:
Changed responsibility from points to speed.

Stan-Kudri committed:
Changed:
-Interface "ISnake": Remove "const char SymbolSnake".
-Class "GameMap": Add static method "CreatSnake".

Stan-Kudri committed:
Changed the condition for selecting the type of snake by pressing.

Stan-Kudri committed:
Changed:
-Changed from "Interface ISnake" to "Abstract class Snake".
-Changed the classes "SnakeNotPassingBorders" and "SnakePassingBorders" under the abstract class "Snake".

blowin commented on Snake/ComponentsGame/GameMap.cs:
```diff
@@ -13,7 +13,7 @@ public class GameMap
         public event Action<Food>? OnEatScore;
 
         private readonly Border _border;
-        private readonly ISnake _snake;
```
пусть будет интерфейс, просто от него наследуй абстрактный класс

blowin commented on Snake/ComponentsGame/GameMap.cs:
```diff
@@ -22,18 +22,7 @@ public GameMap(int width, int height, TypeSnake type, int snakeLength = 10)
             _border = new Border(width, height);
```
передавай Border в конструктор и змейку

blowin commented on Snake/ComponentsGame/GameMap.cs:
```diff
@@ -97,5 +86,22 @@ public void Clear()
             return null;
         }
+
+        private static Snake? CreateSnake(TypeSnake type, Border border, int snakeLength = 10)
```
это не ответственность этого класса

blowin replied:
почему карта сама создаёт какой-то тип змейки, логичнее передать нужный тип на карту

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/Snake.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public void Clear() => _body.ForEach(x => x.Clear());
+
+        public abstract bool Intersect();
```
если основная логика и смысл это для этого метода, то лучше выделить эту логику в отдельную сущность и пусть он принимает змейку(и ещё что-то, если надо) и проверяет необходимое, змейка останеться не тронутой

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/Snake.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public bool IntersectBody(Point food) => _body.Contains(food);
+
+        private protected abstract Point NextPoint();
```
аналогично с вынесенеием метода, змейка вообще не причём к генерации точек

blowin commented on Snake/ComponentsGame/Speed.cs:
```diff
@@ -2,16 +2,29 @@
         public int Value { get; private set; }
 
-        public void Increase() => Value = Value - ValueIncreaseSpeed > 0 ? Value - ValueIncreaseSpeed : Value;
+        public void Increase(int score)
```
удаляй от сюда, есть отдельная ветка с этим

Stan-Kudri committed:
Add interface "ISnake".

Stan-Kudri committed:
Remove commits "Changed responsibility from points to speed."

Stan-Kudri committed:
Changed:

-Class "GameMap":
   1)Changed the input parameters of the constructor;
   2)Removed static method "CreateSnake".

-Extension class/methods ("DirectionsExtension", "FoodExtension" and "PointExtension") add empty line.

-Add Extension method "CreateTypeSnakeGame".

-Class "Game":Changed the input parameters of the constructor.

Stan-Kudri committed:
Rename method from "NextPoint" to "GetNewHeadPosition".

Stan-Kudri committed:
Changed:
-Class "GameMap": Remove emty line in constructor.
-Enum "Directions": Add empty line.
-Extension class "FoodExtension": Add empty line.
-Changed extension methods in class "PointExtension".
-Changed class "Point".

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/ISnake.cs:
```diff
@@ -0,0 +1,21 @@
+
+        public void Clear();
+
+        public abstract bool Intersect();
```
зачем тут abstract?

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/Snake.cs:
```diff
@@ -27,16 +27,16 @@ public Snake(int x, int y, int length)
         public bool EatFood(Point food)
         {
-            if (food.Equals(NextPoint()))
+            if (food.Equals(GetNewHeadPosition()))
```
почему не сохранить результат и использовать его?

blowin commented on Snake/ComponentsGame/Score.cs:
```diff
@@ -30,7 +34,11 @@ public void Draw()
         {
             Points += food.Score;
-            OnUpIntervalScore?.Invoke(Points);
+            if (Points == (_numberInterval + 1) * _speedUpInterval)
```
опять.. ты мержил? если да, то зачем? если нет, то почему не вернул?

blowin commented on Snake/ComponentsGame/Speed.cs:
```diff
@@ -2,29 +2,16 @@
     public class Speed
     {
-        public const int ValueIncreaseSpeed = 25;
+        public const int ValueIncreaseSpeed = 20;
```
уже другая скорость.. не понимаю что происходит

blowin commented on Snake/Extension/SnakeTypeExtension.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public static class SnakeTypeExtension
+    {
+        public static Snake? CreateTypeSnakeGame(this TypeSnake type, Border border, int snakeLength = 10)
```
CreateSnake

blowin commented on Snake/Extension/SnakeTypeExtension.cs:
```diff
@@ -0,0 +1,25 @@
+            }
+            else
+            {
+                return null;
```
лучше Exception, ты его всё равно потом выбрасываешь

Stan-Kudri committed:
Changed:

-Class "Snake":
  1)The "_head" field takes the value in the "EatFood" method;
  2)Remove line "_head = GetNewHeadPosition()", because the "_head" field is assigned in the "EatFood" method.

-Changed methods "Intersect" in classes "SnakeNotPassingBorders" and "SnakePassingBorders".

-Changed method "CreatSnake" in class "SnakeTypeExtension".

Stan-Kudri committed:
Deleted all branch changes "Add snake speed." and fixed the bug.

Stan-Kudri committed:
Changed:
-Class "Border": Add property (get) "List<Point> Borders".
-Class "SnakeNotPassingBorders":
  1)Remove field "_heightField" and "_widthField";
  2)Add field "List<Point> _border";
  3)Remove methods "IsEncounterTheBorder";
  4)Add methods border collision checks "ObstacleCollision";
  5)Changed methods "Intersect".

Stan-Kudri committed:
Changed the game to one type, but the transition through the fields occurs if there are no obstacles.

Stan-Kudri committed:
Remove empty line.

Stan-Kudri committed:
Add line breaks to the "if".

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/Snake.cs:
```diff
@@ -1,42 +1,45 @@
 namespace GameSnake.ComponentsGame.ItemGameMap.SnakeType
 {
-    public abstract class Snake : ISnake
+    public class Snake : ISnake
```
ISnake тоже можно удалить

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/Snake.cs:
```diff
@@ -1,42 +1,45 @@
         public bool EatFood(Point food)
         {
-            if (food.Equals(GetNewHeadPosition()))
+            _head = GetNewHeadPosition();
```
избавляемся от GetNewHeadPosition (Обсуждали)

blowin commented on Snake/ComponentsGame/ItemGameMap/SnakeType/Snake.cs:
```diff
@@ -50,13 +53,64 @@ public bool EatFood(Point food)
-        public abstract bool Intersect();
+        public bool Intersect()
+        {
+            if (ObstacleCollision())
```
лучше отдельный метод сделать, или назвать метод Intersect как-то иначе, а то не очевидно что он столько всего проверяет

Stan-Kudri committed:
Changed:
-Remove interface "ISnake".
-Class "Snake":Broke the method into two that I used method "Intersect".

Stan-Kudri committed:
Add empty line and moved below properties.

Stan-Kudri committed:
Merge branch 'master' into TwoTypesGame
